cli/openApiSpec.ts

```diff
@@ -0,0 +1,276 @@
+
+type FunctionType = 'Action' | 'Mutation' | 'Query' | 'HttpAction'
+
+export type FunctionSpec = {
+    url: string,
+    functions: AnalyzedFunction[]
+};
+
+export type AnalyzedFunction = {
+    identifier: string
+    functionType: FunctionType
+    visibility: Visibility
+    args: ValidatorJSON | null
+    returns: ValidatorJSON | null
+};
+
+function generateSchemaFromValidator(validatorJson: ValidatorJSON): string {
+    switch (validatorJson.type) {
+        case 'null':
+            // Necessary because null only becomes explicitly supported in OpenAPI 3.1.0
+            return 'type: string\nnullable: true'
+        case 'number':
+            return 'type: number'
+        case 'bigint':
+            throw new Error('bigint unsupported')
+        case 'boolean':
+            return 'type: boolean'
+        case 'string':
+            return 'type: string'
+        case 'bytes':
+            throw new Error('bytes unsupported')
+        case 'any':
+            return '{}'
+        case 'literal':
+            if (typeof validatorJson.value === 'string') {
+                return `type: string\nenum:\n  - "${validatorJson.value}"` as string
+            } else if (typeof validatorJson.value === 'boolean') {
+                return `type: boolean\nenum:\n  - ${validatorJson.value.toString()}`
+            } else {
+                return `type: number\nenum:\n  - ${validatorJson.value!.toString()}`
+            }
+        case 'id':
+            return `type: string\ndescription: ID from table "${validatorJson.tableName}"`
+        case 'array':
+            return `type: array\nitems:\n${reindent(
+                generateSchemaFromValidator(validatorJson.value),
+                1
+            )}`
+        case 'record':
+            return 'type: object'
+        case 'object': {
+            const requiredProperties: string[] = []
+            const members: string[] = Object.entries(validatorJson.value).map(
+                ([key, value]) => {
+                    if (!value.optional) {
+                        requiredProperties.push(key)
+                    }
+                    return `${key}:\n${reindent(
+                        generateSchemaFromValidator(value.fieldType),
+                        1
+                    )}`
+                }
+            )
+            const requiredPropertiesStr =
+                requiredProperties.length === 0
+                    ? ''
+                    : `required:\n${reindent(
+                        requiredProperties.map((r) => `- ${r}`).join('\n'),
+                        1
+                    )}\n`
+            const propertiesStr =
+                members.length === 0
+                    ? ''
+                    : `properties:\n${reindent(members.join('\n'), 1)}`
+            return `type: object\n${requiredPropertiesStr}${propertiesStr}`
+        }
+        case 'union': {
+            const nullMember = validatorJson.value.find((v) => v.type === 'null')
```

Comment:
if this is code I wrote, I wouldn't be surprised if there were some bugs around nulls in nested unions. might be worth testing a few of these manually, or we can address them as they come up

Reply:
okay, i can explicitly test this out more. i spent most of today converting `golinks` to use a generated clients, i confirmed that most things work correctly.